File: fastlane/ruby/warnings_markdown_generator.rb
require_relative 'issue_fetcher'

class WarningsMarkdownGenerator
  SWIFT_PACKAGES = "Third Party Swift Packages Warnings"
  SDK_AND_CHAT_WARNINGS = "SDK and MEGAChatSDK Warnings"
  ANALYTICS_LIBRARY = "Analytics Library Warnings"
  OTHER_WARNINGS = "Other Warnings"
  SWIFT_PACKAGES_PATH = "SwiftPackages"
  SDK_PATH = "Modules/DataSource"
  ANALYTICS_PATH = "MEGAAnalyticsiOS"

  def generate_markdown(warnings_json_path, output_file_path)
    warnings = fetch_all_issues(warnings_json_path, :warning)
    markdown = initialize_markdown_sections
    markdown = append_warnings_to_markdown(warnings, markdown)
    markdown = append_close_tags(markdown)

    header = "Warnings\n---"
    markdown_text = header + build_final_markdown(markdown)
    
    File.write(output_file_path, markdown_text) unless markdown_text == header
    markdown_text != header
  end

  private def initialize_markdown_sections
    { 
      swift_packages: initial_markdown_text(SWIFT_PACKAGES),
      sdk_and_chat_warnings: initial_markdown_text(SDK_AND_CHAT_WARNINGS),
      analytics_library: initial_markdown_text(ANALYTICS_LIBRARY),
      other_warnings: initial_markdown_text(OTHER_WARNINGS)
    }
  end

  private def append_warnings_to_markdown(warnings, markdown) 
    warnings.each do |warning|
      file = warning[:file]
      reason = warning[:reason]
      line_number = warning[:line]
      filename = file.empty? ? "" : File.basename(file)

      current_directory = Pathname.getwd
      row_text = "| ⚠️ | #{reason} | #{filename} | #{line_number} |\n"

      case file
      when /#{SWIFT_PACKAGES_PATH}/
        markdown[:swift_packages] += row_text unless markdown[:swift_packages].include?(row_text)
      when /#{SDK_PATH}/
        markdown[:sdk_and_chat_warnings] += row_text unless markdown[:sdk_and_chat_warnings].include?(row_text)
      when /#{ANALYTICS_PATH}/
        markdown[:analytics_library] += row_text unless markdown[:analytics_library].include?(row_text)
      else
        markdown[:other_warnings] += row_text unless markdown[:other_warnings].include?(row_text)
      end
    end

    markdown
  end

  private def append_close_tags(markdown)
    markdown.each_key { |key| markdown[key] += "</details>" }
    markdown
  end

  private def initial_markdown_text(summary_name, is_open = false)
    open_attribute = is_open ? "open" : ""
  
    <<~END
      <details #{open_attribute}>
      <summary>#{summary_name}</summary>
    
      | | Warnings | File | Line |
      | :--- | :--- | :--- | :--- |
    END
  end

  private def build_final_markdown(markdown)
    final_markdown = ""

    markdown.each do |category, content|
      if has_non_header_rows?(content)
        final_markdown += "\n\n" + content
      end
    end

    final_markdown
  end  

  private def has_non_header_rows?(content)
      content_lines = content.split("\n")
      content_lines.length > 6 # Minimum number of lines to include non-header rows
  end

end

File: scripts/WarningParsingKit/Sources/WarningParsingKit/WarningsMarkdownGenerator.swift
import Foundation

final class WarningsMarkdownGenerator {
    struct Issue {
        let file: String
        let reason: String
        let line: String
    }

    enum IssueKind {
        case warning
    }

    // MARK: - Constants
    private enum Constants {
        static let swiftPackages = "Third Party Swift Packages Warnings"
        static let sdkAndChatWarnings = "SDK and MEGAChatSDK Warnings"
        static let analyticsLibrary = "Analytics Library Warnings"
        static let otherWarnings = "Other Warnings"
        
        static let swiftPackagesPath = "SwiftPackages"
        static let sdkPath = "Modules/DataSource"
        static let analyticsPath = "MEGAAnalyticsiOS"
    }

    // MARK: - Public API
    @discardableResult
    func generateMarkdown(warnings: [Issue], outputFilePath: String) throws -> Bool {
        var markdown = initializeMarkdownSections()
        markdown = appendWarningsToMarkdown(warnings, markdown: markdown)
        markdown = appendCloseTags(markdown)

        let header = "Warnings\n---"
        let markdownText = header + buildFinalMarkdown(markdown)

        if markdownText != header {
            try ensureDirectoryExists(for: outputFilePath)
            try markdownText.write(toFile: outputFilePath, atomically: true, encoding: .utf8)
            return true
        }

        return false
    }

    // MARK: - Sections
    private enum SectionKey: CaseIterable {
        case swiftPackages
        case sdkAndChatWarnings
        case analyticsLibrary
        case otherWarnings
    }

    private func initializeMarkdownSections() -> [SectionKey: String] {
        return [
            .swiftPackages: initialMarkdownText(summaryName: Constants.swiftPackages, isOpen: false),
            .sdkAndChatWarnings: initialMarkdownText(summaryName: Constants.sdkAndChatWarnings, isOpen: false),
            .analyticsLibrary: initialMarkdownText(summaryName: Constants.analyticsLibrary, isOpen: false),
            .otherWarnings: initialMarkdownText(summaryName: Constants.otherWarnings, isOpen: true)
        ]
    }

    // MARK: - Append warnings
    private func appendWarningsToMarkdown(_ warnings: [Issue], markdown: [SectionKey: String]) -> [SectionKey: String] {
        var updated = markdown

        for warning in warnings {
            let file = warning.file
            let reason = warning.reason
            let lineNumber = warning.line

            let filename: String
            if let lastPathComponent = URL(string: file)?.lastPathComponent {
                filename = lastPathComponent
            } else {
                filename = ""
            }

            let rowText = "| ⚠️ | \(reason) | \(filename) | \(lineNumber) |\n"

            // Match Ruby case/when with regex-ish substring checks
            if file.contains(Constants.swiftPackagesPath) {
                appendUnique(rowText, to: .swiftPackages, in: &updated)
            } else if file.contains(Constants.sdkPath) {
                appendUnique(rowText, to: .sdkAndChatWarnings, in: &updated)
            } else if file.contains(Constants.analyticsPath) {
                appendUnique(rowText, to: .analyticsLibrary, in: &updated)
            } else {
                appendUnique(rowText, to: .otherWarnings, in: &updated)
            }
        }

        return updated
    }

    private func appendUnique(_ rowText: String, to key: SectionKey, in markdown: inout [SectionKey: String]) {
        guard var content = markdown[key] else { return }
        if !content.contains(rowText) {
            content += rowText
            markdown[key] = content
        }
    }

    // MARK: - Close tags
    private func appendCloseTags(_ markdown: [SectionKey: String]) -> [SectionKey: String] {
        var updated = markdown
        for key in SectionKey.allCases {
            if var content = updated[key] {
                content += "</details>"
                updated[key] = content
            }
        }
        return updated
    }

    // MARK: - Templates
    private func initialMarkdownText(summaryName: String, isOpen: Bool) -> String {
        let openAttribute = isOpen ? "open" : ""
        return """
        <details \(openAttribute)>
        <summary>\(summaryName)</summary>

        | | Warnings | File | Line |
        | :--- | :--- | :--- | :--- |
        
        """
    }

    // MARK: - Final markdown build
    private func buildFinalMarkdown(_ markdown: [SectionKey: String]) -> String {
        var finalMarkdown = ""

        let orderedKeys: [SectionKey] = [.swiftPackages, .sdkAndChatWarnings, .analyticsLibrary, .otherWarnings]

        for key in orderedKeys {
            guard let content = markdown[key] else { continue }
            if hasNonHeaderRows(content) {
                finalMarkdown += "\n\n" + content
            }
        }

        return finalMarkdown
    }

    private func hasNonHeaderRows(_ content: String) -> Bool {
        let lines = content.split(separator: "\n", omittingEmptySubsequences: false)
        return lines.count > 6
    }

    private func ensureDirectoryExists(for filePath: String) throws {
        let url = URL(fileURLWithPath: filePath)
        let directoryURL = url.deletingLastPathComponent()

        let fileManager = FileManager.default
        if !fileManager.fileExists(atPath: directoryURL.path) {
            try fileManager.createDirectory(
                at: directoryURL,
                withIntermediateDirectories: true,
                attributes: nil
            )
        }
    }
}
